language/bytecode-verifier/src/binary_views.rs

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::control_flow_graph::VMControlFlowGraph;
+use libra_types::vm_status::StatusCode;
+use move_core_types::{
+    account_address::AccountAddress, identifier::IdentStr, language_storage::ModuleId,
+    vm_status::VMStatus,
+};
+use vm::{
+    access::{ModuleAccess, ScriptAccess},
+    errors::VMResult,
+    file_format::{
+        AddressIdentifierIndex, CodeUnit, CompiledScript, Constant, ConstantPoolIndex, FieldHandle,
+        FieldHandleIndex, FieldInstantiation, FieldInstantiationIndex, FunctionDefinition,
+        FunctionDefinitionIndex, FunctionHandle, FunctionHandleIndex, FunctionInstantiation,
+        FunctionInstantiationIndex, IdentifierIndex, Kind, ModuleHandle, ModuleHandleIndex,
+        Signature, SignatureIndex, SignatureToken, StructDefInstantiation,
+        StructDefInstantiationIndex, StructDefinition, StructDefinitionIndex, StructHandle,
+        StructHandleIndex,
+    },
+    CompiledModule,
+};
+
+// A `BinaryIndexedView` provides table indexed access for both `CompiledModule` and
+// `CompiledScript`.
+// Operations that are not allowed for `CompiledScript` return an error.
+// A typical use of a `BinaryIndexedView` is while resolving indexes in bytecodes.
+pub(crate) enum BinaryIndexedView<'a> {
```

Comment:
Hooray enums!

I feel we lean on traits way way too much when we want "interface" or "object" like behavior. Enums behave just like parent/child classes, and it feels like the right application here